### PR TITLE
EPE-165: Improve logic for unlinkable blocks while sync'ing

### PR DESCRIPTION
**Pull request by rusty-block-one**
The current logic for handling unlinkable blocks while sync'ing closes
the node prematurely when the node is flooded with unlinkable blocks in
a short time window. This fix groups unlinkable blocks into time windows
in order to manage the flooding condition.

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [x] Stability bug fix

- [ ] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:1965]: This needs to be tracked per connection.

Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:1965]: The def_max_consecutive_rejected_blocks should be part of block_status_monitor

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:139]: This is 2 seconds not 2ms

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:144]: explicit

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:1566]: This is extremely unlikely. Not sure I'm following the logic here.
accepted() should restart this monitor always.

**Review by heifner — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/net_plugin.cpp
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:1579]: My thinking is that for consecutive windows, the end of one window is not necessarily the start of the next window.  To illustrate:
t0, the start of the first time window
te0, the time the first window closes
t1, the start of the next window
s, the window size
t, the time of the rejected block that closes the first window

te0 = t0 + s
t != te0
t1 = t instead of te0

The window_start is set to "0" in accepted().  However, the rejected() logic uses a window_start=0 as a signal to set the start of the next window.  My thinking is that this ensures each time window has at least 1 rejected block. My concern is that if the interval between rejected blocks is > than a multiple of the window size (s), a rejected block could result in the count being increased by more than one. This may be overthinking the problem.

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:2015]: This can be outside the mutex. Move up.

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:1976]: This mutex can be moved down to right above `sync_last_requested_num`

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:1579]: I think this overcomplicates it. Seems to me it should just be if you are in same window then don't increment; otherwise increment.

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:553]: explicit

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:1568]: Seems to me if you are starting a new window you should increment events_.
I think it would be clearer to use `if( events_ == 0 ) {` here.

**Review by heifner — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/net_plugin.cpp
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp]: Changed logic to make it clearer

File: plugins/net_plugin/net_plugin.cpp
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:1560]: Not that performance is a real issue here, but this is likely to be faster to just set the values. Although the compiler has probably removed this if anyway. I think it is clearer to just set the values. But not a huge deal either way.

**Review by heifner — APPROVED** [review]

**Comment by heifner** [comment]
Approved, but please don't merge until LRT has been run on this change. All net_plugin changes should have LRTs run before merge.

File: plugins/net_plugin/net_plugin.cpp
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:1560]: I appreciate your comment and thinking.  I debated internally the difference between just resetting the values and testing the current state. I'll change it to just setting the values. 
My internal debate centered around wondering what the most likely state on each call.  If the most likely state is accepted then the if test would reduce cpu cycles with a hiccup on when the state changed.  If the state is continually going back and forth, then removing the if test would most likely reduce cpu cycles.

**Pull request merged**
